plugin/vscode/performance-writer.js

```diff
@@ -0,0 +1,113 @@
+// Copyright (C) 2020  Matthew Glazar
+// See end of file for extended copyright information.
+
+const fs = require("fs");
+const path = require("path");
+const process = require("process");
+
+class PerformanceWriter {
+  constructor(extensionContext) {
+    this._extensionContext = extensionContext;
+    this._dataAttributes = {
+      columns: ["timestamp", "file-ref", "event-id", "file-name-or-duration"],
+    };
+    this._tmpLogLines = [];
+    this._fileIndex = 0;
+    this._writeToFileTimer = null;
+    this._timerActivated = false;
+    if (!fs.existsSync(this._extensionContext.logUri.fsPath)) {
+      fs.mkdirSync(this._extensionContext.logUri.fsPath);
+      this.writeHeaderLine();
+    }
+    this._performanceTxtPath = path.join(
+      this._extensionContext.logUri.fsPath,
+      "qljsPerformanceTrace.txt"
+    );
+    this._writableStream = fs.createWriteStream(this._performanceTxtPath, {
+      encoding: "utf8",
+      flags: "a",
+    });
+    this._currentSessionMap = {};
+  }
+
+  fileNameOptimizer(fileName) {
+    if (
+      Object.prototype.hasOwnProperty.call(this._currentSessionMap, fileName)
+    ) {
+      this._fileIndex = this._currentSessionMap[fileName];
+    } else {
+      this._fileIndex++;
+      this._currentSessionMap[fileName] = this._fileIndex;
+      this.writeNewFileLine(fileName);
+    }
+  }
+
+  initiateLogging(timeTaken, fileName) {
+    this.fileNameOptimizer(fileName);
+    this.writeExistingFileLine(timeTaken);
+    this.setWriteToFileTimer();
+  }
+
+  writeHeaderLine() {
+    this._tmpLogLines.push(...this._dataAttributes.columns);
+  }
+
+  writeNewFileLine(fileName) {
+    this._tmpLogLines.push(Date.now(), this._fileIndex, "n", fileName);
+  }
+
+  writeExistingFileLine(timeTaken) {
+    this._tmpLogLines.push(Date.now(), this._fileIndex, "p", timeTaken);
+  }
+
+  setWriteToFileTimer() {
+    if (!this._timerActivated) {
+      this._timerActivated = true;
+      this._writeToFileTimer = setInterval(() => {
+        this._armTimer();
+      }, 2000);
+    }
+  }
+
+  _armTimer() {
+    if (this._tmpLogLines.length === 0) {
+      this._timerActivated = false;
+      clearInterval(this._writeToFileTimer);
+    } else {
+      this.writeToFile();
+    }
+  }
+
+  writeToFile() {
+    while (this._tmpLogLines.length > 0) {
+      let tracingLine = this._tmpLogLines.splice(0, 4);
+      if (typeof tracingLine[0] === "string") {
+        tracingLine = tracingLine.join(",");
+      } else {
+        tracingLine = "\n" + tracingLine.join(",");
+      }
```

Comment:
Let's always add the newline prefix.

Let's say that, for whatever reason, we had a PerformanceWriter for this file which crashed in the middle of writing. The file might look like this (without a trailing newline):
```
timestamp,file-ref,event-id,file-name-or-duration
1234,1,
```
If we start up a PerformanceWriter, then we'll write the header again, but without a leading newline. The file will end up looking like this:
```
timestamp,file-ref,event-id,file-name-or-duration
1234,1,timestamp,file-ref,event-id,file-name-or-duration
4567,1,n,banana.js
4567,1,p,12473
```
The file is corrupted.

Reply:
I realize now that you only write the header line if the file didn't already exist. So I guess the situation I mentioned won't cause corruption.

Also, you mentioned in chat that the file wouldn't be reused in practice anyway.